python/feldera/client.py

```diff
@@ -0,0 +1,398 @@
+from typing import Optional
+import logging
+import time
+
+from feldera.config import Config
+from feldera.connector import Connector
+from feldera.attached_connector import AttachedConnector
+from feldera.program import Program
+from feldera.pipeline import Pipeline
+from feldera._httprequests import HttpRequests
+
+
+class Client:
+    """
+    A client for the Feldera HTTP API
+
+    A client instance is needed for every Feldera API method to know the location of
+    Feldera and its permissions.
+    """
+
+    def __init__(
+            self,
+            url: str,
+            api_key: Optional[str] = None,
+            timeout: Optional[int] = None,
+    ) -> None:
+        """
+        :param url: The url to Feldera API (ex: https://try.feldera.com)
+        :param api_key: The optional API key for Feldera
+        :param timeout: (optional) The amount of time in seconds that the cient will wait for a response beforing timing
+            out.
+        """
+
+        self.config = Config(url, api_key, timeout)
+        self.http = HttpRequests(self.config)
+
+    def programs(self) -> list[Program]:
+        """
+        Get all programs
+        """
+        resp = self.http.get(
+            path="/programs",
+        )
+
+        return [
+            Program(
+                name=program.get("name"),
+                program=program.get("code"),
+                description=program.get("description"),
+            ) for program in resp
+        ]
+
+    def get_program(self, name: str, with_code: bool = False) -> Program:
+        """
+        Get a program by name
+        :param name: The name of the program
+        :param with_code: If True, the program code will be included in the response
+        """
+        resp = self.http.get(
+            path=f"/programs/{name}",
+            params={
+                "with_code": with_code,
+            }
+        )
+
+        return Program(
+            name=resp.get("name"),
+            program=resp.get("code"),
+            description=resp.get("description"),
+            status=resp.get("status"),
+            version=resp.get("version"),
+        )
+
+    def compile_program(self, program: Program):
+        """
+        Compiles a SQL program
+        :param program: The program to compile

```

Comment:
i would say that this may block for a long time and on error it throws
we may eventually need to separate validating sql from compiling Rust

Reply:
Yes, @ryzhyk and I have talked about this and plan on using the validate api once we have it.